src/core/sock/sockinfo_tcp.cpp

```diff
@@ -5258,6 +5195,92 @@ int sockinfo_tcp::rx_wait_helper(int &poll_count, bool blocking)
     return ret;
 }
 
+bool sockinfo_tcp::poll_and_process_rx(bool *is_timeout)
+{
+    uint64_t poll_sn = 0;
+
+    // There's only one CQ
+    m_rx_ring_map_lock.lock();
+
+    // We need to consider what to do in case poll_and_process_element_rx fails on try_lock.
+    // It can be too expansive for the application to get nothing just because of lock contention.
+    // In this case it will be better to have a lock() version of poll_and_process_element_rx.
+    // And then we should continue polling untill we have ready packets or we drained the CQ.
+    auto prev_sndbuf = sndbuf_available();
+    bool all_drained = true;
+    if (likely(m_p_rx_ring)) {
+        all_drained = m_p_rx_ring->poll_and_process_element_rx(&poll_sn);
+    } else { // There is more than one CQ, go over each one
+        rx_ring_map_t::iterator rx_ring_iter;
+        for (rx_ring_iter = m_rx_ring_map.begin(); rx_ring_iter != m_rx_ring_map.end();
+             rx_ring_iter++) {
+            if (unlikely(rx_ring_iter->second->refcnt <= 0)) {
+                __log_err("Attempt to poll illegal cq");
+                continue;
+            }
+
+            all_drained &= rx_ring_iter->first->poll_and_process_element_rx(&poll_sn);
+        }
+    }
+    m_rx_ring_map_lock.unlock();
+    lock_tcp_con(); // We must take a lock before checking m_n_rx_pkt_ready_list_count
+
+    bool sndbuf_change = (sndbuf_available() != prev_sndbuf);
+    if (likely(m_n_rx_pkt_ready_list_count || !all_drained || sndbuf_change ||
+               m_ready_conn_cnt)) { // Got completions from CQ
+        __log_entry_funcall("Ready %d packets. sn=%llu", m_n_rx_pkt_ready_list_count,
+                            (unsigned long long)poll_sn);
+        IF_STATS(m_p_socket_stats->counters.n_rx_poll_hit++);
+        unlock_tcp_con();
+        return true;
+    }
+
+    IF_STATS(m_p_socket_stats->counters.n_rx_poll_miss++);
+    if (is_timeout) {
+        *is_timeout = m_loops_timer.is_timeout(); // We do this under lock.
+    } else { // Case when we poll after request notification and before going to sleep.
+        // We do it here already under lock to avoid another lock/unlock calls.
+        m_sock_wakeup_pipe.going_to_sleep();
```

Comment:
AFAIU, we need to "go to sleep" before the `request_notification()`. Looks like the condition is wrong - after is_timeout is assigned by pointer, we should go to sleep if `!is_timeout || !*is_timeout`.

Besides, calling `going_to_sleep()` here hides this execution flow where it is not expected. This is not critical for now, but for the future, it will be better to call it before request_notification() / epoll_wait() when decision to sleep is made.

Reply:
I didn't change the time of calling to going_to_sleep(). afaik it does not matter the order between request_notification() and going_to_sleep. It matters to call going_to_sleep() before actual wait.
I refactored this PR splitting to more commits and more clear changes.